backend/index.js

```diff
@@ -44,71 +56,133 @@ io.on("connection", (socket) => {
     socket.to(data.room).emit("receive_message", data);
   });
 
-	socket.on("send_image", (data) => {
-		console.log("Image received from", data.author, data.room);
+  socket.on("send_image", (data) => {
+    console.log("Image received from", data.author, data.room);
     socket.to(data.room).emit("receive_image", data);
     console.log("Image sent to", data.room);
-	});
+  });
 
   socket.on("disconnect", () => {
     console.log("A user has left the platform");
   });
 });
 
+// Custom Auth Middleware
+const authorization = (req, res, next) => {
+  const token = req.cookies.access_token;
+  if (!token) {
+    return res
+      .status(403)
+      .json({ message: "Proper authentication not supplied" });
+  }
+  try {
+    const data = jwt.verify(token, JWT_SECRET);
+    req.room = data.room;
+    req.room = data.role;
+    return next();
+  } catch {
+    return res.status(403);
+  }
+}
+
 // REST Endpoints
 app.get('/', (req, res) => {
-  res.json({
-    rooms: roomStore
-  })
+  return res
+    .status(200)
+    .json({
+      rooms: roomStore
+    });
 });
 
 app.get('/login', async (req, res) => {
   const username = req.query.username;
   const password = req.query.password;
+
+  const token = jwt.sign({ room: username, role: 'admin' }, JWT_SECRET);
   await client.get(username).then(storedPassword => {
     if (storedPassword == null) {
-      res.json({
-        authenticated: false,
-        reason: "No user found"
-      });
+      res
+        .status(403)
+        .json({
+          authenticated: false,
+          reason: "No user found"
+        });
       return;
     }
     if (storedPassword == password) {
-      res.json({
-        authenticated: true
-      });
-      return;
+      res
+        .cookie("access_token", token, {
+          secure: process.env.NODE_ENV === 'production',
+          sameSite: 'none'
+        })
+        .json({
+          authenticated: true
+        })
+        .status(200);
+        return;
     } else {
       res.json({
         authenticated: false,
         reason: "Incorrect password"
-      });
+      }).status(403);
       return;
     }
   });
 });
 
-// app.post('/register', async (req, res) => {
-//   const username = req.body.username;
-//   const password = req.body.password;
-//   if (await client.exists(username)) {
-//     res.sendStatus(403);
-//   } else {
-//     await client.set(username, password);
-//     res.sendStatus(200);
-//   }
-// });
-
-app.post('/:room', (req, res) => {
+// verification endpoint for JWT tokens
+app.post("/verifyRoom", authorization, (req, res) => {
+  const token = req.cookies.access_token;
+  const room = req.query.room;
+  const data = jwt.verify(token, JWT_SECRET);
+  const token_role = data.role;
+  const token_room = data.room;
+  console.log(data, token_role, "attempting to join room", room, "with", token_room);
+  if (token_room === room || token_role === 'admin') {
```

Comment:
we should implement some additional form of authentication for the admin role, some kind of rate limitation so that in the event an admin account is stolen, it can't be used to view all open rooms at once. I suggested adding a second array at line 35 so we can track logged in admin users, and we can add additional functionality to prevent an admin from viewing more than 2 or 3 rooms at once.

Reply:
This is a good idea, lets make an issue!